### ISSUE TITLE
Wait for translate to be loaded before displaying

### ISSUE THREAD

**Issue by JerryBels** [angular-translate/angular-translate#1344]
Hello people,

I want to wait for the translation to be ready before displaying the page - to avoid the bad UX where the text is flickering when first launching the app.

I was using the filter, like that : 

    {{'tabs.gallery' | translate}}

And transformed it to use the service instead : 

	$translate(['tabs.wall', 'tabs.gallery']).then(function (translations) {
		$scope.tabsWall = translations['tabs.wall'];
		$scope.tabsGallery = translations['tabs.gallery'];
	});

It didn't work, and I understood why by looking at the [docs](http://angular-translate.github.io/docs/#/guide/03_using-translate-service) : 

    $translate service works asynchronously, which means it returns the translation for the given translation id, as soon as it could determine it. If it doesn't exist it'll probably just return the translation id.

So I also tried to use the directive, didn't help. I added `$translate.onReady` before using the service, didn't work. I tried to use `translate-cloak`, didn't work. Wanted to use the `resolve` of the `ui router`'s state... Didn't find a way to do that efficiently either. I could potentially use in that resolve : 

	$rootScope.$on('$translateChangeSuccess', function () {
		deferred.resolve(true);
	});

BUT then of course it works only on the first load of the page, and it seems very inaccurate to me.

Sooooo any help ?

Thanks !

**Comment by knalli** [comment]
Can you play with http://plnkr.co/edit/tF8QqyMukl4XKinixKZP?p=info ? This is a 2.9 pre release containing fixes.

**Comment by JerryBels** [comment]
Hello, and thanks for answering,

which method should I try with that new release ? Filter ? Directive ? Service ? Something else ?

**Comment by JerryBels** [comment]
Okay, I tried with many ways, didn't help :(

**Comment by knalli** [comment]
> 
> which method should I try with that new release ? Filter ? Directive ? Service ? Something else ?

The translate-cloak directive.

**Comment by mr-White** [comment]
Watching eagerly. 

The flickering of translatable ID's before the JSON dictionaries resolve are an unfortunate side-effect of what appears to be the design of angular-translate's $translate service. 

I believe it would yield a better UX if the service were to return blank strings instead of translatable ID's. That way simple implementations of the app with async loading of JSON dictionaries are not coupled with a flash of translation ID's.

If the service is mean't to return translation ID's to help the dev debug their app, then I hope this is changed since dev's should rely on the useMissingTranslationHandlerLog() for recording missing translation ID's

tldr; I'm reading this thread with great interest. Ping me, if there is a way I can help

**Comment by JerryBels** [comment]
@mr-White +1, great idea to return blank instead of the key, I think (or at least give this possibility as a configuration option).

**Comment by knalli** [comment]
@mr-White  @JerryBels  Are you both aware of http://angular-translate.github.io/docs/#/guide/17_custom-error-handler ? This should make the "blank return on error" solution possible for you.

**Comment by knalli** [comment]
> Something I didn't specified, all my files are in local (no remote loading) so I really don't get why the filter (or any other way, whatever) returns the key first before the translation takes place and sets the right text. Since there IS a value to display ( the key, for example 'tabs.gallery' ), even ng-cloak won't work...

If you translations are local (you mean defined with `translationsProvider.translations()`, the `translate-cloak` will be useless for you. Or say in other words: I should be not possible getting a filter returning translation ids which are configured "statically". Can you please share this configuration?

**Comment by mr-White** [comment]
@knalli Thank you for the share! 

@tinyfly check out the custom error handler for dealing with the FUOC, peace ~

**Comment by JerryBels** [comment]
Hey, thanks for the share, I will take a look at what I can do with it :)

As for my configuration, I will detail my setup a bit more. So it's a mobile app based on Ionic framework - which relies on Angular. 

I have a file "locales.js` included in `index.html` which contains : 

	var translations = {
		languages: [
			'en',
			'fr'
		]
	};

In my `.config` I'm using `$translateProvider` to configure the languages as follow : 

    for (var i=0; i<translations.languages.length; i++)
                $translateProvider.translations(translations.languages[i], translations[translations.languages[i]]);

            $translateProvider.fallbackLanguage('en');
            $translateProvider.useSanitizeValueStrategy('escape');

And that's all. After that the routing hapens, and on the loaded page I first see the translation ID, no matter if I use the directive, filter or service to get the translations, and then the wanted translation appears.

**Comment by albertytau** [comment]
Both ng-cloak, translate-cloak are not working for my ionic app.
Within the menu, part of the elements are translated and part are not translated.
Please help

**Comment by tspaeth** [comment]
@albertytau Without any example code - how should we help?

**Comment by albertytau** [comment]
Thanks @tspaeth, i do run my code on ionic as well. My code like what @JerryBels presented.
It runs perfectly on a slow device which can display after the translation function completed. However, when I run on my iPhone 6s, partial labels are translated and partial labels are not translated. It may because I have a long list of labels for translation. I try to sort my labels and move those not transalted labels on the top of the list, but it does not work. How can I ensure all labels are translated before it can be displayed. Interestingly. it switches the translation if I press on the label which is a side menu item.

**Comment by tspaeth** [comment]
It's still too general from your description... but are you aware of the FOUC - chapter? http://angular-translate.github.io/docs/#/guide/12_asynchronous-loading or the cloak directive?

**Comment by knalli** [comment]
Please provide a working demo showing the issue. Otherwise we cannot solve this.

**Comment by JerryBels** [comment]
Hey @knalli ,

a demo will be hard to provide since it's only seeable on phones and using Cordova, a codepen or such opened with a regular PC would look good. When testing my app on PC there is no issue.

I could provide a video of what is happening, along with code, if it can help.

Also I still have to try the custom error handler that you mentionned before, just didn't have time to work on this again yet (rush at work, you guys know how it is), will do this in a week or two.

**Comment by knalli** [comment]
Well. Running in a Cordova webapp container has less loading timeout than a regular webpage.. just because everything is local. I won't say there is no loading timeout, but the issue will be more visible on a regular web page setup.

Said that, you can run a "Cordova app" on plnkr or so -- because it is only a webapp running locally. It's limited to browser APIs, but that not an issue. Anything Cordova related will not be part of this one here anyway.

@JerryBels  @albertytau  Without knowing about your configurations (you have still don't show us anything), we cannot help you. The `translate-cloak` is working, but only with its specific limitations. The cloak works until the first language is loaded -- or the *specific* key is present (being resolved).

Also, just a few marks: 

# If you see filter strings (i.e. `{{'TITLE' | translate}}`) in the output, this will be **uncompiled AngularJS expressions**. angular-translate cannot handle this because AJS itself has not compiled it yet. This would be an issue of your template, you app structure. Perhaps a missing ng-cloak or a missing $digest. Or a non-AJS-plugin maybe?
# If you see the translate ID (i.e. `TITLE` instead of `Title`), this will be a **compiled AJS expression** and therefor is eventually under control of angular-translate. The usage of translate-cloak should help.

Perhaps you should give the browsers dev tools a try also: For example Chrome offers tools for bandwidth control which give you the possibility to test under "bad connection" situations.

----

Unless you can show as a working case, we can not fix this issue. And without more information, we cannot help.

**Comment by rdesoky** [comment]
@JerryBels, make sure that you call $translate() after $translate.use(/*lang_id*/) returned promise has been resolved.

**Comment by JerryBels** [comment]
@rdesoky thanks, man ! You saved the day. 

Here was the issue : I was calling a function that did `$translate.use(/lang_id/)`, but in the `deviceready` event callback ! Calling it out of this callback fixed the issue.

**Comment by rdesoky** [comment]
@JerryBels I am glad that I could help
